### PR TITLE
Allow recursive bindings with _

### PR DESCRIPTION
**Pull request by OlivierNicole**
## Description of the problem from the [Compiler hacking wiki](https://github.com/ocamllabs/compiler-hacking/wiki/Things-to-work-on):

Currently, you cannot use _ in recursive bindings:

``` ocaml
let rec _ = () and x = 3;;
Characters 8-9:
  let rec _ = () and x = 3;;
          ^
Error: Only variables are allowed as left-hand side of `let rec'
```

This is because patterns aren't allowed there, only variables. Really _ should be considered a variable in this context.
## What has been done

Just like with `let` bindings, `_` is now allowed as a variable name in `let rec` bindings. As with `let` bindings, this construction is translated into a sequence.

A test has also been added to the test suite to make sure that `_` bindings work and can produce effects. Evaluation order is not tested as it may be changed by flambda.

### CODE REVIEW

File: testsuite/tests/letrec/underscore_binding.ml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+(*           Gabriel Scherer, projet Gallium, INRIA Rocquencourt          *)
+(*                                                                        *)
+(*   Copyright 2016 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
```
Comment (damiendoligez) [inline testsuite/tests/letrec/underscore_binding.ml]: Please don't use headers in testsuite files. Especially headers with the wrong names...

File: testsuite/tests/letrec/underscore_binding.ml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+let p = print_endline
+
+(* Test "_" bindings in `let rec` constructs. All bindings produce the same
+ * effect to avoid testing evaluation order. *)
+let test =
+  let rec _ = p "1"; p "2"
+      and x = p "1\n2"; 3
```
Comment (yallop) [inline testsuite/tests/letrec/underscore_binding.ml]: It'd be good to have the bindings used recursively in the test -- i.e. use `x` in the rhs of `y` and vice versa, and also use them both in the rhs of the `_` bindings.

**Comment by yallop** [comment]
Could you please add an entry to [`Changes`](https://github.com/ocaml/ocaml/blob/4aac28049299591032842a7781f750f9ff1783c0/Changes#L5)?

**Comment by OlivierNicole** [comment]
After modifying the test as you suggested, it makes my code crash because of an unknown identifier, because of the way I translated "_"-bindings into a sequence. I am working on the problem. So this PR is faulty, sorry about that.

**Pull request closed without merging**